textlock/textlock.1.in
.TH textlock 1 "@RELEASE_DATE@" "@PACKAGE_STRING@" "User Commands"
.SH NAME
textlock \- password-lock a terminal
.SH SYNOPSIS
\fBtextlock\fP [ \fB-h\fP ]

\fBtlpasswd\fP
.SH AVAILABILITY
This program will not run on an ASCII terminal which is smaller than 40
columns by 15 rows. 
.SH DESCRIPTION
The \fBtextlock\fP utility locks an ASCII terminal until the user who
invoked it enters their password at the keyboard. While the terminal
is locked, a clock is displayed at random locations on the screen. If
a key is pressed, the program displays a dialog box in which the user
is expected to enter their password. If an invalid password is
entered, or if no input is received after a certain amount of time,
the program reverts back to the clock display. The program cannot be
interrupted with any of the keyboard-generated interrupts (for
example, control-C, break, control-Z, and control-Y).

By default, the user's system password will unlock the
display. However, the user may specify an alternate password to use
for unlocking the display by running the \fBtlpasswd\fP program. This
program asks for a new password and stores an encrypted version of the
password in the file \fI.textlock\fP in the user's home
directory. \fBtextlock\fP reverts to verifying against the user's
system password only if the \fI.textlock\fP file does not exist.
.SH OPTIONS
Without any options, \fBtextlock\fP immediately locks the screen. With
the \fI-h\fP option, it displays usage and credits
information.

Several key commands are available in the password editing box. These are
summarized below (`^' denotes the `CONTROL' key).
.TP 5
.B Backspace, Delete, or Left-Arrow
Any of these keys will delete the character to the left of the cursor.
.TP 5
.B ^U or Clear-Line
These keys will clear the entire editing line and reposition the cursor at
the beginning.
.TP 5
.B ^L or Refresh
These keys will cause the entire screen to be refreshed. This option is
useful if a password entry is interrupted by a broadcast message.
.TP 5
.B Return or Enter
These keys will accept the current entry.
.TP 5
.B Space-Bar
This key will abort the input and return to the clock display.
.TP 5
.B ^T
This is an escape code that specifies that the following keycode
should be treated literally. This is useful for entering passwords
that contain embedded control characters. Any character may be entered
after a ^T, including another ^T.

.TP 0
The \fBtlpasswd\fP utility accepts no options.
.SH NOTES
Due to the security concerns posed by a user entering his or her password
into a running program, \fBtextlock\fP takes some precautions about
handling this information. A static buffer is used to hold the user's
password input, and this buffer is erased immediately after the
input is encrypted for comparison against the actual encrypted
password.

Another security problem would arise if \fBtextlock\fP were to dump
core immediately after the user entered their password, since the core
file (process image) could be examined to find the password. To
prevent this from happening, \fBtextlock\fP reroutes every signal
which could cause a core dump to a special signal handler. This
handler zeroes the input buffer and then exits gracefully. In most
cases this will prevent a core dump altogether.
.SH FILES
.PD 0
.TP 20
.B /etc/passwd
to get password and user information

.PD 0
.TP 20
.B ~/.textlock
to get encrypted textlock password

.SH SEE ALSO
.B xlock(1)

.SH BUGS
The \fBtlpasswd\fP program may not properly handle the input of
control characters and escape sequences as part of a
password. Additionally, only the first 8 characters of the password
are significant.

.SH AUTHOR
Mark Lindner <@PACKAGE_BUGREPORT@>


ChangeLog
2017-10-28  Mark Lindner  <markl@avalon>

	* stat/stat.c - fixed compiler warning
	* watcherd/* - removed
	* pfortune/* - removed
	* configure.ac - bumped version

2012-10-01  Mark Lindner  <markl@neuromancer>

	* aux-build/* - updated files
	* pkgenv/pkgenv.c - fixed compiler warnings

2011-06-26  Mark Lindner  <markl@neuromancer>

	* bin2c/* - added new utility
	* configure.ac, Makefile.am - added new bin2c utility, bump version

2008-04-23  Mark Lindner  <markl@neuromancer>

	* pfortune.1.in - fix typo in manpage

2008-04-21  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - bugfix for shell handling

2008-04-20  Mark Lindner  <markl@neuromancer>

	* texi2ps, texi2ps.1.in - new script
	* pkgenv.c - bugfix: implement proper version string comparator
	* configure.ac - bump version

2007-12-10  Mark Lindner  <markl@neuromancer>

	* man2ps, man2ps.1.in, ps2pdf, ps2pdf.1.in, texi2pdf, texi2pdf.1.in,
	  man2pdf, man2pdf.1.in - modified scripts to accept multiple filename
	  arguments
	* <multiple> - rename 'tformat' utility to 'wrap'
	* pkgenv.c - fix memory crash bug

2007-12-08  Mark Lindner  <markl@neuromancer>

	* ranline.c - bugfix

2007-12-06  Mark Lindner  <markl@neuromancer>

	* ranline.c, ranline.1.in - add support for reading from stdin;
	  increase maximum supported line length to 1024

2007-11-19  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - added -V (verbose) flag
	* basecvt.c, basecvt.1.in - added '?' command

2007-11-11  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - improve error messages
	* pkgenv/automake_patch.txt - added

2007-09-16  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - allow environment variable references in
	  variable substitution

2007-09-05  Mark Lindner  <markl@neuromancer>

	* pkgenv.c, pkgenv.1.in - more bugfixes & improvements

2007-09-01  Mark Lindner  <markl@neuromancer>

	* pkgenv.c - add special logic to handle old vs. new MANPATH dir;
	  don't add dirs that don't exist; bugfixes to 'del' command;
	  add DYLD_LIBRARY_PATH support for Mac OS X
	* *.1, *.1.in, configure.ac - generate manpages with correct version &
	  date in header
	* scripts/man2pdf - new script
	* Makefile.am - add 'pdf' target

2007-08-26  Mark Lindner  <markl@neuromancer>

	* <multiple> - code cleanup, bugfixes
	* pkgenv/, basecvt/ - new utilities added

2007-02-08  Mark Lindner  <markl@neuromancer>

	* scripts/Makefile.am - add manpages to dist
	* pfortune.c - fix segmentation fault

2006-04-17  Mark Lindner  <markl@neuromancer>

	* scripts/ - added useful shell scripts

2006-04-17  Mark Lindner  <markl@neuromancer>

	* dirstack.c, textlock.c, configure.ac - check for both ncurses.h
	  and ncurses/ncurses.h; fixes build error on OS X.

2006-04-15  Mark Lindner  <markl@neuromancer>

	* pfortune.c, pfortune.1 - partial rewrite, new -l switch
	* pfortune - added new databases (smileys, paul_graham, fortunes)

2006-04-13  Mark Lindner  <markl@neuromancer>

	* <multiple> - modified build files to locate libcfl using pkg-config
	* smileys.fdb - new fortune database added

2005-08-15  Mark Lindner  <markl@neuromancer>

	* dirstack.1, pascii.1 - manpage corrections
	* README - updated

2005-08-12  Mark Lindner  <markl@neuromancer>

	* <multiple> - Merged in 'djinn' project.
	* dirstack.c - Truncate paths that do not fit on one line.

2004-05-12  Mark Lindner  <markl@neuromancer>

	* bat.c - Added -8 switch.
	* <multiple> - Various minor code fixes.

2003-02-22  Mark Lindner  <markl@neuromancer>

	* <multiple> - Reorganized package so that there is only one
	  configure.ac (at the top level)
	* <multiple> - Modified all source code to use CFL library (which
	  replaces the pingutil library)
	* stat.c - don't exit with an error if getgrnam() or getpwent() fails,
	  but rather just display "?" for the user or group name
	* dirstack.c - fixed bug that created repaint glitch on delete,
	  added handling of ^L for screen refresh, and added -p switch.
	* ranline.c - use random() instead of rand() so that (on Solaris, at
	  least) we're not limited to a random number range of 32767.
	* <multiple> - updated manpages
	* pascii.c - handle DEL key and high-bit chars more gracefully
	* textlock.c - use C_system_get_hostname() to get hostname; the old way
	  of copying the results of uname() was non-portable, because SYS_NMLN
	  is not defined on some systems (namely OS X).

2003-01-02  Mark Lindner  <markl@neuromancer>

	* <multiple> - replaced `bool_t' with `p_bool_t' to fix compilation
	  errors when linked against the latest pingutil

2002-04-17  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - reorganized the entire source tree to use automake &
	  autoconf. Damn this was a lot of work (especially since I'd never
	  used automake before). Also made several minor fixes to make the
	  packages compile cleanly on Linux.

2002-04-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* dirstack - new utility

2001-10-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - minor modifications for pingutil-2.2
	* watcherd - new utility
	* pfortune - new utility

2001-09-14  Mark Lindner  <mark_a_lindner@yahoo.com>

	* <multiple> - slight modifications to use pingutil-2.1
	* ptree - renamed to dirtree to avoid confusion with Solaris ptree cmd
	* dirtree - fixed segmentation fault


textlock/textlock.c
/* ----------------------------------------------------------------------------
   textlock - text console lock utility
   Copyright (C) 1994-2011  Mark A Lindner

   This file is part of misctools.

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see
   <http://www.gnu.org/licenses/>.
   ----------------------------------------------------------------------------
*/

/* --- Feature Test Switches --- */

#include "config.h"

/* --- System Headers --- */

#include <stdio.h>
#include <sys/types.h>
#include <pwd.h>
#include <sys/param.h>
#include <time.h>
#ifdef HAVE_SYS_TIME_H
#include <sys/time.h>
#endif /* HAVE_SYS_TIME_H */
#include <signal.h>
#if defined HAVE_NCURSES_NCURSES_H
#include <ncurses/ncurses.h>
#elif defined HAVE_NCURSES_H
#include <ncurses.h>
#endif
#include <termios.h>

#include <cbase/cbase.h>

/* --- Local Headers --- */

#include "textlock.h"

/* --- Macros --- */

/* --- File Scope Variables --- */

static char *pass, inbuf[TL_MAXPASS+1], text1[40], text2[40],
  *id = TL_SHORTID, *inst = TL_INSTR;
static struct termios tio_new, tio_old;
static int cattr = 0;
static WINDOW *mainwin;

/* --- Functions --- */

static void tl_init(void);
static void tl_quit(void);
static c_bool_t tl_input(void);
static void tl_sleep(void);
static c_bool_t tl_verify(char *);
static void tl_crash(int);

int main(int argc, char **argv)
{
  C_error_init(*argv);

  if(argc == 2)
    if(!strcmp(argv[1], "-h"))
    {
      C_error_printf("%s\n", TL_ID);
      C_error_usage(TL_USAGE);
      exit(EXIT_SUCCESS);
    }  

  tl_init();

  for(;;)
  {
    halfdelay(TL_TIMEOUT);
    tl_sleep();
    if((tl_input()))
    {
      if(tl_verify(inbuf))
        break;
      beep(), beep();
    }
  }

  tl_quit();
  return(EXIT_SUCCESS);
}

/*
 *  tl_verify: verify entered password
 *
 *  Encrypt the password _s_ and compare it to the encrypted password
 *  from the passwd file. Return TRUE if it's a match.
 */

static c_bool_t tl_verify(char *s)
{
  c_bool_t ok = C_system_passwd_validate(s, pass);

  C_zeroa(s, TL_MAXPASS, char);

  return(ok);
}

/*
 *  tl_sleep: wait for the input of a password
 *
 *  Draw a floating clock on the text screen while waiting for input.
 *  Return when a key is pressed.
 */

static void tl_sleep(void)
{
  time_t t1;
  struct tm *t2;
  chtype c;
  char clock[TL_CLOCKSZ+1];
  int x, y, i;

  flushinp(); /* flush typeahead */

  for(;;)
  {
    /* draw the clock */

    time(&t1);
    t2 = localtime(&t1);
    strftime(clock, TL_CLOCKSZ, "[ %l:%M %p ]", t2);

    /* find a random spot to put it & draw it there */

    y = C_random(LINES);
    x = C_random(COLS - TL_CLOCKSZ);

    cattr = 1 - cattr;
    if(cattr)
      wattron(mainwin, A_REVERSE);
    mvwaddstr(mainwin, y, x, clock);
    wattroff(mainwin, A_REVERSE);
    wrefresh(mainwin);

    /* get a keypress */

    c = mvwgetch(mainwin, 0, 0);

    /* erase the clock */

    wmove(mainwin, y, x);
    for(i = 0; i < TL_CLOCKSZ; waddch(mainwin, ' '), i++);
    wrefresh(mainwin);

    if(c != ERR)
      break;
  }

  /* got a keypress, so jump out */

  return;
}

static int tl_centerstr(char *s)
{
  return((COLS - strlen(s)) / 2);
}

/*
 *  tl_input: input a password
 *
 *  Input the password in a dialog box. Return TRUE if we got a
 *  string, or FALSE if there was a timeout.
 */

static c_bool_t tl_input(void)
{
  int xpos, i, xb = (COLS - (TL_MAXPASS + 4)) / 2, yb = (LINES - 3) / 2;
  chtype c;
  char *p;
  c_bool_t flag = FALSE, literal = FALSE;

  flushinp(); /* flush typeahead */

  /* draw the centered titles */

  mvwaddstr(mainwin, 1, tl_centerstr(text1), text1);
  mvwaddstr(mainwin, 2, tl_centerstr(text2), text2);
  mvwaddstr(mainwin, LINES-2, tl_centerstr(id), id);
  mvwaddstr(mainwin, LINES-4, tl_centerstr(inst), inst);

  /* draw box for input field */

  for(i = 1; i < (TL_MAXPASS + 4); i++)
  {
    mvwaddch(mainwin, yb, xb + i, '-');
    mvwaddch(mainwin, yb + 2, xb + i, '-');
  }

  mvwaddch(mainwin, yb + 1, xb, '|');
  mvwaddch(mainwin, yb + 1, xb + TL_MAXPASS + 4, '|');

LOOP:
  p = inbuf;
  *p = 0;
  xpos = 0;
  literal = FALSE;

  wrefresh(mainwin);

  /* input loop with delete capabilities */

  halfdelay(TL_INPUT_TIMEOUT);
  for(;;)
  {
    c = mvwgetch(mainwin, yb + 1, xb + xpos + 2);

    if(c == ERR)
      break;

    *p = c;

    if((xpos < TL_MAXPASS) && (literal || (*p > ' ' && *p <= '~')))
      literal = FALSE, wechochar(mainwin, '*'), xpos++, p++;

    else if(*p == ' ')
      break;

    else if(((*p == KEY_BS) || (c == KEY_DEL) || (c == KEY_LEFT)
             || (c == KEY_DC)) && xpos)
      xpos--, p--, wechochar(mainwin, '\b'), wechochar(mainwin, ' '),
        literal = FALSE;

    else if((*p == '\f') || (c == KEY_REFRESH))
      wrefresh(curscr), literal = FALSE;

    else if((*p == KEY_CR) || (c == KEY_ENTER))
    {
      *p = 0;
      flag = TRUE;
      break;
    }

    else if((*p == KEY_CTRLU) || (c == KEY_EOL))
    {
      for(i = 0; i < TL_MAXPASS; i++)
        mvwaddch(mainwin, yb+1, xb+2+i, ' ');
      goto LOOP;
    }

    else if((*p == KEY_CTRLT))
      literal = TRUE;

    else beep();
  }

  wclear(mainwin);

  return(flag);
}

/*
 *  tl_init: initialization function
 *
 *  Trap all signals; get user's login name, finger name, password and
 *  salt, host name; initialize screen. Exit if any of the operations fail.
 */

static void tl_init(void)
{
  char *host, *name, *user;
  struct passwd *pw;

  /* trap all keyboard-generated signals */

  signal(SIGINT, SIG_IGN);
  signal(SIGQUIT, SIG_IGN);
  signal(SIGTERM, SIG_IGN);
  signal(SIGTSTP, SIG_IGN);

  /* trap all signals that could cause a core dump */

  signal(SIGSEGV, tl_crash);
  signal(SIGBUS, tl_crash);
  signal(SIGFPE, tl_crash);
  signal(SIGILL, tl_crash);
  signal(SIGTRAP, tl_crash);
  signal(SIGIOT, tl_crash);
#ifdef SIGEMT
  signal(SIGEMT, tl_crash);
#endif
  signal(SIGSYS, tl_crash);

  if(!(pw = getpwuid(getuid())))
  {
    C_error_printf("Can't find your password entry.\n");
    exit(EXIT_FAILURE);
  }

  if(strlen(pw->pw_passwd) < 3)
  {
    FILE *fp;
    char text[80];

    if(! C_system_cdhome())
    {
      C_error_printf("Unable to access home directory.\n");
      exit(EXIT_FAILURE);
    }

    if(!(fp = fopen(".textlock", "r")))
    {
      C_error_printf("Can't open ~/.textlock file.\n");
      exit(EXIT_FAILURE);
    }

    C_io_gets(fp, text, sizeof(text), '\n');
    fclose(fp);

    pass = C_string_dup(text);
  }
  else
    pass = C_string_dup(pw->pw_passwd);

  user = C_system_get_login();
  name = C_system_get_fullname();
  host = C_system_get_hostname();

  /* turn off XON/XOFF processing (that is, don't allow freeze of
     screen output with control-S */

  tcgetattr(STDIN_FILENO, &tio_old);
  tio_new = tio_old;

  tio_new.c_iflag &= ~(IXON | IXOFF);
  tio_new.c_lflag &= ~(ECHO | ICANON | ISIG | IEXTEN);

  tio_new.c_cc[VINTR] =
    tio_new.c_cc[VSTART] =
    tio_new.c_cc[VSTOP] =
    tio_new.c_cc[VQUIT] =
    tio_new.c_cc[VERASE] =
    tio_new.c_cc[VKILL] =
    tio_new.c_cc[VEOF] =
    tio_new.c_cc[VEOL] =
#ifdef SOLARIS
    tio_new.c_cc[VSWTCH] =
#endif
    -1;

  tcsetattr(STDIN_FILENO, TCSANOW, &tio_new);

  mainwin = initscr();

  if((COLS < 40) || (LINES < 15))
  {
    endwin();
    C_error_printf("Can't run in a window less than 40x15.\n");
    exit(EXIT_FAILURE);
  }

  noecho();
  halfdelay(TL_TIMEOUT);
  keypad(mainwin, TRUE);

  C_random_seed();

  sprintf(text1, "Locked by %s", name);
  sprintf(text2, "(%s@%s)", user, host);
}

/*
 */

static void tl_crash(int sig)
{
  /* zap the input buffer in case of a core dump */

  memset(inbuf, 0, TL_MAXPASS);
  C_error_printf("Aborted due to signal #%i\n", sig);
  exit(EXIT_FAILURE);
}

/*
 */

static void tl_quit(void)
{
  wclear(mainwin);
  wrefresh(mainwin);
  endwin();

  /* reset old terminal attributes */

  tcsetattr(STDIN_FILENO, TCSANOW, &tio_old);
  exit(EXIT_SUCCESS);
}

/* end of source file */
